addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/httpfuzzer/HttpFuzzerHandler.java

```diff
@@ -147,7 +146,9 @@ private HttpFuzzer createFuzzer(
 
         MessageLocationReplacer<HttpMessage> replacer =
                 MessageLocationReplacers.getInstance()
-                        .getMLR(HttpMessage.class, TextHttpMessageLocation.class);
+                        .getMLR(
+                                HttpMessage.class,
+                                fuzzLocations.get(0).getMessageLocation().getClass());
```

Comment:
Hi @thc202 ,

I found that this has breaked the existing fuzzer reason being that ```fuzzLocations.get(0).getMessageLocation().getClass()``` returns the concrete instance class i.e. in case of HttpFuzzer DefaultHttpTextMessageLocation but HttpFuzzer registers the  TextHttpMessageLocationReplacerFactory which returns target class as interface and hence while getting replacer code fails.

please suggets.

Reply:
Did the above change to fix the issue. Please suggest.

Reply:
It's preferable (more extensible) to change `MessageLocationReplacers` to walk the class hierarchy of the message location when getting the factory, e.g.:
```diff
--- a/addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/messagelocations/MessageLocationReplacers.java
+++ b/addOns/fuzz/src/main/java/org/zaproxy/zap/extension/fuzz/messagelocations/MessageLocationReplacers.java
@@ -84,10 +84,35 @@ public class MessageLocationReplacers {
         if (replacers != null) {
             @SuppressWarnings("unchecked")
             MessageLocationReplacerFactory<T> replacerFactory =
-                    (MessageLocationReplacerFactory<T>) replacers.get(messageLocationClass);
+                    (MessageLocationReplacerFactory<T>) getFactory(replacers, messageLocationClass);
             return replacerFactory.createReplacer();
         }
 
         return null;
     }
+
+    private static <T> T getFactory(
+            Map<Class<? extends MessageLocation>, T> factories, Class<?> clazz) {
+        if (!MessageLocation.class.isAssignableFrom(clazz)) {
+            return null;
+        }
+
+        T factory = factories.get(clazz);
+        if (factory != null) {
+            return factory;
+        }
+
+        factory = getFactory(factories, clazz.getSuperclass());
+        if (factory != null) {
+            return factory;
+        }
+
+        for (Class<?> interfaceClazz : clazz.getInterfaces()) {
+            factory = getFactory(factories, interfaceClazz);
+            if (factory != null) {
+                return factory;
+            }
+        }
+        return null;
+    }
 }
```

Reply:
hmm I was also thinking about it but then i looked at the method name and it is ``` getTargetMessageLocation```
and from this it seems like it should be the target class. But makes sense if we do this way then there are no changes required in HttpFuzzer.

thanks,
Karan

Reply:
And it is, `TextHttpMessageLocationReplacerFactory` is targeting `TextHttpMessageLocation` (an interface), so it should handle any implementation of said location.

Do you mean `HttpFuzzerHandler`? (There are no changes in `HttpFuzzer`.)

Reply:
Yes yes, i mean HttpFuzzerHandler

Reply:
That change is still needed though, the one in `ExtensionHttpFuzzer` would no longer be.

Reply:
oh sorry, i got the point. just misplaced the names.